CONTRIBUTING.md

```diff
@@ -40,177 +264,782 @@ since someone else can often do this if you include a complete example. We
+partial results much more quickly (when our CI is not overloaded).
+
+The main documentation resources on our CI are:
+
+- the [README for users, i.e. plugin and library authors][CI-README-users];
+- the [README for developers, and contributors][CI-README-developers];
+- the README of the [user-overlays][] directory.
+
+Preparing an overlay is a step that everyone goes through at some
+point.  All you need to know to prepare an overlay manually is in the
+README in the [user-overlays][] directory.  You might
+want to use some additional tooling such as the `make ci-*` targets of
+`Makefile.ci`, the Nix support for getting the dependencies of the
+external projects (see the README in [`dev/ci/nix`](dev-ci-nix), and
+the (undocumented so far) `dev/tools/create_overlays.sh` script.
+
+More work is to be done on understanding how each developer proceeds
+to prepare overlays, and propose a simplified and documented
+procedure.
+
+We also have a benchmarking infrastructure, which is documented [on
+the wiki][jenkins-doc].
+
+#### Code owners, issue and pull request templates ####
+
+These files can be found in the [`.github`](.github) directory.  The
+templates are particularly useful to remind contributors what
+information we need for them, and, in the case of PRs, to update the
+documentation, changelog, and test-suite when relevant.
+
+GitHub now supports setting up multiple issue templates, and we could
+use this to define distinct requirements for various kind of bugs,
+enhancement and feature requests.
+
+#### Style guide ####
+
+There exists an [old style guide][old-style-guide] whose content is
+still mostly relevant.  We don't use a code formatter at the current
+time, and we are reluctant to formatting-only commits, or commits
+formatting parts of code that are unchanged beyond formatting.
+However, it is still a good idea if you don't know how to format a
+block to use the formatting that [ocamlformat][] would give
+
+#### OCaml resources ####
+
+You may find lots of OCaml resources on <http://ocaml.org/>, including
+documentation, a Discourse forum, the package archive, etc.  You may
+also want to refer to the [Dune documentation][dune-doc].
+
+#### Git documentation, tips and tricks ####
+
+Lots of resources about git, the version control system, are available
+on the web, starting with the [official website][git].
+
+We recommend a setup with two configured remotes, one for the official
+Coq repository, called `upstream`, and one for your fork, called
+`origin`.  Here is a way to do this for a clean clone:
+
+``` shell
+git clone https://github.com/coq/coq -o upstream
+cd coq
+git remote add origin git@github.com:$YOURNAME/coq
+# Make sure you click the fork button on GitHub so that this repository exists
+cp dev/tools/pre-commit .git/hooks/ # Setup the pre-commit hook
+```
+
+Then, if you want to prepare a fix:
+
+``` shell
+git checkout master
+git pull # Make sure we start from an up-to-date master
```

Comment:
We should say to use `-ff-only`: when it fails it will indicate that the user did some work on their master branch without making a mess.

Reply:
Maybe even tell people to put `pull.ff=only` in their git config